src/bootstrap/setup.rs

```diff
@@ -110,15 +110,50 @@ pub fn setup(src_path: &Path, profile: Profile) {
 
     let build = TargetSelection::from_user(&env!("BUILD_TRIPLE"));
     let stage_path = ["build", build.rustc_target_arg(), "stage1"].join("/");
-    let toolchain_link =
-        Command::new("rustup").args(&["toolchain", "link", "stage1", &stage_path[..]]).output();
-    if toolchain_link.is_err() {
-        println!("WARNING: `rustup`failed to link stage 1 build to `stage1` toolchain");
-        println!(
-            "To manually link stage 1 build to `stage1` tool chain, run:\n
-            `rustup toolchain link stage1 build/{}/stage1`",
-            build.rustc_target_arg()
-        );
+
+    let created = fs::create_dir(&stage_path[..]).is_ok();
+
+    // Link `stage1` toolchain if `rustup` is installed
+    // and we have not already done so.
+    if Path::new("~/.cargo/bin/rustup").exists() && created {
+        // Let's see if we have already linked the stage1 toolchain.
+        let toolchain_list = Command::new("rustup")
+            .args(&["toolchain", "list"])
+            .stdout(std::process::Stdio::piped())
+            .output();
+
+        if toolchain_list.is_ok() {
+            let toolchain_list = toolchain_list.unwrap();
+            let links = String::from_utf8_lossy(&toolchain_list.stdout);
+
+            if !links.contains("stage1") {
+                let toolchain_link = Command::new("rustup")
+                    .args(&["toolchain", "link", "stage1", &stage_path[..]])
+                    .output();
+                if toolchain_link.is_err() || !toolchain_link.unwrap().status.success() {
+                    println!("`rustup` failed to link stage 1 build to `stage1` toolchain");
+                    println!(
+                        "To manually link stage 1 build to `stage1` toolchain, run:\n
+                        `rustup toolchain link stage1 {}`",
+                        &stage_path[..]
+                    );
+                } else {
+                    println!(
+                        "Added `stage1` rustup toolchain; try `cargo +stage1 build` on a separate rust project to run a newly-built toolchain"
+                    );
+                }
+            } else {
+                println!(
+                    "`stage1` toolchain already linked; `rustup` will not attempt to link `stage1` toolchain",
+                );
+            }
+        } else {
+            println!(
+                "`rustup` failed to list the current toolchains; `rustup` will not attempt to link `stage1` toolchain"
```

Comment:
I don't think saying "rustup" is correct here - it's x.py that's making the decision. But the output comes from x.py already so I don't think we need to mention that explicitly.
```suggestion
                "`rustup` failed to list the current toolchains; not attempting to link `stage1` toolchain"
```